mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -2224,17 +2022,39 @@ static func::FuncOp createGpuGemmKernel(ModuleOp module,
   SmallVector<NamedAttribute, 2> funcAttrs = {
       b.getNamedAttr("kernel", b.getUnitAttr()),
       b.getNamedAttr("mhal.arch", archAttr)};
+  SmallVector<Type, 3> flatTypes =
+      llvm::map_to_vector(argTypes, flattenShapedType);
   auto func =
       b.create<func::FuncOp>(loc, isVerifier ? kernelNameVerifier : kernelName,
-                             b.getFunctionType(argTypes, {}), funcAttrs);
+                             b.getFunctionType(flatTypes, {}), funcAttrs);
   if (reverse_grid) {
     func->setAttr(rock::ReverseGridAttrAttr::getMnemonic(), b.getUnitAttr());
   }
 
+  SmallVector<StringRef> aNames = {"g", "m", "k"}, bNames = {"g", "k", "n"},
+                         cNames = {"g", "m", "n"};
+  if (transposeA)
+    std::swap(aNames[1], aNames[2]);
+  if (transposeB)
+    std::swap(bNames[1], bNames[2]);
+  if (transposeC)
+    std::swap(cNames[1], cNames[2]);
+
   Block *block = func.addEntryBlock();
   b.setInsertionPointToStart(block);
-  Value aVal = block->getArgument(0), bVal = block->getArgument(1),
-        cVal = block->getArgument(2);
+  SmallVector<Value, 3> expandedArgs;
+  expandedArgs.reserve(3);
+
+  ArrayRef<StringRef> dimLists[3] = {aNames, bNames, cNames};
+  for (auto [arg, t, names] :
+       llvm::zip(block->getArguments(), argTypes, dimLists)) {
+    auto logicalType = cast<MemRefType>(t);
+    rock::TopDownTMBuilder flattener(b, names, logicalType.getShape(), loc);
+    flattener.unmerge("raw", 0, names, logicalType.getShape());
+    rock::TransformMapAttr expandAttr = flattener.get();
+    expandedArgs.push_back(b.create<rock::TransformOp>(loc, arg, expandAttr));
```

Comment:
I see similar code to this used by conv2d generator and in another place in rocmlir-gen. I am wondering if there is a way to avoid duplication.

Also, I think this is kind of crucial to this PR (basically you have 1D memrefs and you provide a logical view that gets flattened, IIUC), so having this code in a function with a bit of comment would make things clearer.

Reply:
I think the means of getting the names are differint enough in the copies of this code that generality is being tricky.

Give ma a thought on the function signature if you've got one?

Reply:
Something on the line of:
```
// Provide a list of 1d memrefs `args`. Each `args[i]` will be reshaped like `types[i]` whose dimension's name will be contained in `names[i]`. The output will be contained in the `logicalViews` output parameter. 
void createLogicalViews(args, types, names, MutableArrayRef logicalViews){

}
```
Would that not be applicable throughout the three places?